llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -5332,11 +5301,31 @@ static SDValue lowerVECTOR_SHUFFLE(SDValue Op, SelectionDAG &DAG,
   if (ShuffleVectorInst::isReverseMask(Mask, NumElts) && V2.isUndef())
     return DAG.getNode(ISD::VECTOR_REVERSE, DL, VT, V1);
 
-  // If this is a deinterleave and we can widen the vector, then we can use
-  // vnsrl to deinterleave.
-  if (SDValue Src =
-          isDeinterleaveShuffle(VT, ContainerVT, V1, V2, Mask, Subtarget))
-    return getDeinterleaveViaVNSRL(DL, VT, Src, Mask[0] == 0, DAG);
+  // If this is a deinterleave(2,4,8) and we can widen the vector, then we can
+  // use shift and truncate to perform the shuffle.
+  // TODO: For Factor=6, we can perform the first step of the deinterleave via
+  // shift-and-trunc reducing total cost for everything except an mf8 result.
+  // TODO: For Factor=4,8, we can do the same when the ratio isn't high enough
+  // to do the entire operation.
+  if (VT.getScalarSizeInBits() < Subtarget.getELen()) {
+    const unsigned MaxFactor = Subtarget.getELen() / VT.getScalarSizeInBits();
+    assert(MaxFactor == 2 || MaxFactor == 4 || MaxFactor == 8);
+    for (unsigned Factor = 2; Factor <= MaxFactor; Factor <<= 1) {
+      unsigned Index = 0;
+      if (ShuffleVectorInst::isDeInterleaveMaskOfFactor(Mask, Factor, Index) &&
+          1 < count_if(Mask, [](int Idx) { return Idx != -1; })) {
+        if (SDValue Src = getSingleShuffleSrc(VT, ContainerVT, V1, V2)) {
+          if (Src.getValueType() == VT) {
+            EVT WideVT = VT.getDoubleNumVectorElementsVT();
```

Comment:
Not directly related to this patch, but I think for large minimum VLEN this might create an illegal type. There's an attempt to protect this in `getSingleShuffleSrc` by checking `RISCVTargetLowering::getLMUL(ContainerVT) != RISCVII::VLMUL::LMUL_8`. For sufficiently large minimum VLEN, LMUL==4 can be the largest legal type and LMUL=8 would be an illegal type because we don't have a large enough MVT to represent it.

Reply:
Oh, nasty.  Do you have a recommended idiom for checking this case?  Example code I can copy from?

Reply:
Do we need to widen here at all? We used to need to widen so that the narrow result type after dividing factor 2 was exactly VT. With larger factors we have to pad back up to VT at the end now. Does  the padding at the end cover this case?

Reply:
Hm, let me play with that.  I ran into a similar-ish case in a follow on change which impacted code quality, so let's see if reworking this slight does too.

Reply:
I made this change, it does work.  It also introduces some additional VL and VTYPE toggles, which is why I think we didn't do this before.

Two questions for you.

1) Are we bothered by the toggles?  I can probably eliminate them < m1 by adding back the removed code, but conditional on LMUL.   For the > m2 cases, this is probably an improvement.  We could potentially treat this a generic improvement to add to e.g. VLOptimizer too.

2) Do we want to separate this change into it's own review?  It feels somewhat borderline to me honestly, so will defer to your preference.

Reply:
> Are we bothered by the toggles? I can probably eliminate them < m1 by adding back the removed code, but conditional on LMUL. For the > m2 cases, this is probably an improvement. We could potentially treat this a generic improvement to add to e.g. VLOptimizer too.

I agree its probably an improvement for the wide LMUL. I don't think I'm bothered by the toggles. Some of the tests are producing undefined elements in the upper half of the shuffle result and then using them in a later instruction. Is that likely in real code?